cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/main.go

```diff
@@ -20,11 +20,12 @@ import (
 	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
 	plugins "github.com/kubeapps/kubeapps/cmd/kubeapps-apis/gen/core/plugins/v1alpha1"
 	v1alpha1 "github.com/kubeapps/kubeapps/cmd/kubeapps-apis/gen/plugins/fluxv2/packages/v1alpha1"
+	"github.com/kubeapps/kubeapps/pkg/kube"
 )
 
 // RegisterWithGRPCServer enables a plugin to register with a gRPC server.
-func RegisterWithGRPCServer(s grpc.ServiceRegistrar) {
-	v1alpha1.RegisterFluxV2PackagesServiceServer(s, NewServer())
+func RegisterWithGRPCServer(s grpc.ServiceRegistrar, config kube.ClustersConfig, unsafeUseDemoSA bool) {
```

Comment:
So, what I'd recommend here is to use the fact that functions are first-class citizens in go, and so can be passed around as parameters. You'll see that I've already used this to be able to test the plugins, but it looks to me like it'd be worth extending that here so that the signature was:

```suggestion
func RegisterWithGRPCServer(s grpc.ServiceRegistrar, clientGetter func(context.Context) (dynamic.Interface, error)) {
```
or more simply by defining the type alias:
```
type DynamicClientGetter func(context.Context) (dynamic.Interface, error)
func RegisterWithGRPCServer(s grpc.ServiceRegistrar, clientGetter DynamicClientGetter) {
```

(see https://golang.org/doc/codewalk/functions/ for more about using functions like this).

That way, we can define the function in the kubeappsapis once, and once only, then pass that function to each plugin when registering. That way, plugins literally don't have to care about the details, they just call the function with the context to get a client (and/or handle the error).

Quite neatly, the only change would then be that the `NewServer(...)` functions would now just accept the client rather than creating it, and we can remove the function for creating clients from both plugins?

Let me know if something doesn't make sense there, or you get stuck. Happy to jump in.

Reply:
> You'll see in ./cmd/root.go that it's sole responsibility is parsing all the args/config, then starting the server by calling server.Serve(...). Currently this has only two args, the port and the plugin dirs, but it is exactly the point where any info required to run the server should go.

> Does the above solve this? ie. if you pass the data you need into the Serve() function, we should have everything we need? My assumption is that that data will be used in the function you create... I'll take a look soon.

Yep, I got it. After refactoring a little bit for writing the tests, I just improved the b) option, creating two functions w/ and w/o the required params (still don't know why golang is against overloading functions...)

> Yeah, though right now you can just update the plugin functions to use the same signature? ...

I copied the example wrongly. I meant `main.DynamicClientGetter` and  `plugin.DynamicClientGetter`. So even if they are the same ` DynamicClientGetter`  since they're coming from different packages and the signatures don't match  :(